src/daemon/rpmostree-sysroot-upgrader.c

```diff
@@ -1189,36 +1105,43 @@ overlay_packages (RpmOstreeSysrootUpgrader *self,
   else
     base_rev = g_strdup (self->new_revision);
 
-  /* create a tmprootfs in which we initially only check out the bare minimum to
-   * make libhif happy. this allows us to provide more immediate feedback to the
-   * user if e.g. resolving/downloading/etc... goes wrong without making them
-   * wait for a full tree checkout first. */
-  if (!checkout_min_tree_in_tmp (repo, base_rev, &tmprootfs, &tmprootfs_dfd,
-                                 cancellable, error))
-    goto out;
+  /* Do we have any layered packages currently, or are we trying to
+   * add some?  If we do, then we need to check out a copy of the tree
+   * so that libdnf can read it to parse the rpmdb, load
+   * /usr/lib/os-release from it, etc.
+   *
+   * If we don't have any layered packages, let's not pay the cost of
+   * checking out another copy of the tree that we'd just toss away
+   * when doing a redeployment.
+   */
+  if (g_strv_length (self->requested_packages) > 0 ||
+      g_hash_table_size (self->packages_to_add) > 0)
+    {
+      if (!checkout_tree_in_tmp (repo, base_rev, devino_cache, &tmprootfs,
+                                 &tmprootfs_dfd, cancellable, error))
+        goto out;
 
-  /* check if there are any items in requested_packages or pkgs_to_add that are
-   * already installed in base_rev */
-  if (!finalize_requested_packages (self, repo, base_rev, tmprootfs_dfd,
-                                    cancellable, error))
-    goto out;
+      /* check if there are any items in requested_packages or pkgs_to_add that are
+       * already installed in base_rev */
+      if (!finalize_requested_packages (self, repo, base_rev, tmprootfs_dfd,
+                                        cancellable, error))
+        goto out;
 
-  /* trivial case: no packages to overlay */
-  if (g_strv_length (self->requested_packages) == 0)
+      if (!overlay_final_pkgset (self, devino_cache, tmprootfs, tmprootfs_dfd,
+                                 repo, base_rev,
+                                 cancellable, error))
```

Comment:
GitHub somehow didn't submit another comment I made here:

So now we may end up calling this even with an empty pkgset. It's probably simpler to just drop the conditional, and keep the trivial case check.